snapfaas/src/vm.rs

```diff
@@ -386,18 +386,32 @@ impl Vm {
     fn send_req(&self, invoke: syscalls::Invoke) -> bool {
         use time::precise_time_ns;
         if let Some(invoke_handle) = self.handle.as_ref().and_then(|h| h.invoke_handle.as_ref()) {
-            let (tx, _) = mpsc::channel();
-            let req = Request {
-                function: invoke.function,
-                payload: serde_json::from_str(invoke.payload.as_str()).expect("json"),
-            };
-            use crate::metrics::RequestTimestamps;
-            let timestamps = RequestTimestamps {
-                at_vmm: precise_time_ns(),
-                request: req.clone(),
-                ..Default::default()
-            };
-            invoke_handle.send(Message::Request((req, tx, timestamps))).is_ok()
+            let invoke_handle = invoke_handle.clone();
+            thread::spawn(move || {
+                let req = Request {
+                    function: invoke.function,
+                    payload: serde_json::from_str(invoke.payload.as_str()).expect("json"),
+                };
+                for _ in 0..5 {
+                    use crate::metrics::RequestTimestamps;
+                    let timestamps = RequestTimestamps {
+                        at_vmm: precise_time_ns(),
+                        request: req.clone(),
+                        ..Default::default()
+                    };
+
+                    let (tx, rx) = mpsc::channel();
+                    if invoke_handle.send(Message::Request((req.clone(), tx, timestamps))).is_err() {
+                        return;
+                    }
```

Comment:
I think this SendError is fatal. `send` errors only when the resource manager side receiver is disconnected.
Plus, we are panicking in multivm/main.rs